NSSL-MP scheme updates

MicroTed opened the pull request:
This NSSL-MP update makes some improvements under the hood. No changes to the public subroutine calls.

  - More accurate saturation mixing ratio calculation (iqvsopt=1)
  - Changed default droplet renucleation to irenuc=5, which allows extra nucleation at high supersaturation
  - A default explicit rain breakup for 3-moment rain (irainbreak=2) has been added to reduce excessive median drop diameters in rain cores, which in turn increases evaporation and cool pool temperature deficits (were too warm).
  - Imposed reflectivity conservation in graupel->hail conversion (ihlcnh=3) and Bigg freezing (both 2- and 3-moment)
 
The same update is being submitted for WRF, as well.

I suppose this will require a new baseline for the regional_wofs regression test.

Edit from @grantfirl:
This PR is also combined with https://github.com/ufs-community/ccpp-physics/pull/277 (documentation changes only)

Output for the SRW v3.0.0 documentation can be found here: https://dtcenter.ucar.edu/GMTB/UFS_SRW_App_v3.0.0/sci_doc/index.html

MicroTed committed:
NSSL-MP scheme update:
  - More accurate saturation mixing ratio calculation (iqvsopt=1)
  - Changed default droplet renucleation to irenuc=5, which allows extra nucleation at high supersaturation
  - A default explicit rain breakup for 3-moment rain (irainbreak=2) has been added to reduce excessive median drop diameters in rain cores, which in turn increases evaporation and cool pool temperature deficits (were too warm).
  - Imposed reflectivity conservation in graupel->hail conversion (ihlcnh=2) and Bigg freezing (both 2- and 3-moment)

MicroTed committed:
Typo fix

grantfirl commented:
@MicroTed I don't see PRs in fv3atm and ufs-weather-model for this. Are you going to add those, or would you like me to?

grantfirl commented:
@MicroTed Also, have you run UFS RTs anywhere yet?

MicroTed commented:
@grantfirl Correct, I have not done the fv3 or ufs PRs yet or run an RT. I think there's a baseline for the wofs regional RT that would need to be updated, too. I don't have any changes for fv3atm other than the ccpp version, so if you have some time to set that up, I would not mind. I just haven't gotten to it yet.

grantfirl committed:
Merge branch 'ufs/dev' into nssl3m-update2025a

MicroTed commented:
@grantfirl For other reasons, I noticed that I need to up the interstitial routine set_aerosol_tracer_index to give the correct result when nssl_3moment = 1. It seems, though, that this routine could be eliminated by using info within the calling subroutines (PBL pre/post). I'll just fix set_aerosol_tracer_index for now, though.

MicroTed committed:
Bug fix for subroutine set_aerosol_tracer_index to set kk correctly when nssl_3moment=true

MicroTed committed:
Merge branch 'nssl3m-update2025a' of https://github.com/MicroTed/ccpp-physics into nssl3m-update2025a

grantfirl committed:
Merge branch 'ufs/dev' into nssl3m-update2025a

grantfirl commented on physics/MP/NSSL/module_mp_nssl_2mom.F90:
```diff
@@ -5929,6 +6078,104 @@ SUBROUTINE calc_eff_radius    &
+            ! Lambda for graupel
+            hwdn = xdn0(lh)
+            IF ( lvh > 1 ) THEN ! variable density
+              IF ( an(ix,jy,kz,lvh) > 1.e-30 ) THEN
```
Suggestion: you may want to define a variable for `1.e-30` in case it ever needs to change.

grantfirl replied:
Same with `1.e-8`

MicroTed replied:
Yeah, easy to do and makes sure things are consistent. I can do it for this or the next update.

grantfirl commented on physics/MP/NSSL/module_mp_nssl_2mom.F90:
```diff
@@ -5929,6 +6078,104 @@ SUBROUTINE calc_eff_radius    &
+              ENDIF
+            ENDIF
+            
+              lam_h = factor_h *((hwdn*cx(mgs,lh))/(qx(mgs,lh)*rho0(mgs)))**(1./3.)
```
`1./3.` is used 125 times in the code. Many schemes define it as a constant at the top of the scheme and use it throughout to achieve a minor speed up. Something to consider.

MicroTed replied:
Yes, I have seen that done. I think the speedup would only be at -O0, since most compilers will precompute the 1./3. with any optimization (maybe even at 0, but I don't know). It's hard to outsmart modern compilers

grantfirl reviewed (approved):
Looks good to me. I left some minor suggestions, but it looks like there are some good code improvements sprinkled in with the science changes. Note to other reviewers: the changes in GFS_PBL_generic*.F90 don't need .meta changes since the added interface variables already existed in the metadata for some reason.

grantfirl commented:
@MicroTed I'm running UFS RTs on Hera with this code as-is right now.

grantfirl commented:
@MicroTed Tests passed except for the expected ones using NSSL MP that changed the results: https://github.com/ufs-community/ufs-weather-model/pull/2661. Note that another test timed out, but that will pass with more wallclock time.

MicroTed commented:
@grantfirl Thank you so much for running these!

MicroTed committed:
module_mp_nssl_2mom.F90 : G. Firl code change suggestion and interface update (unused in CCPP but for consistency with concurrent WRF PR)

MicroTed commented:
>  Note to other reviewers: the changes in GFS_PBL_generic*.F90 don't need .meta changes since the added interface variables already existed in the metadata for some reason.

I assume that's because GFS_PBL_generic_common.F90 doesn't have a .meta file and it just getting called from the pre/post. The set_aerosol_tracer_index routine actually seems to be unnecessary since the value it calculates could be done within the pre/post routines. But it's not a big deal.

I added your suggested code change, which should not change the current results.

grantfirl committed:
bring doc changes only over from scidoc4SRW3 branch

grantfirl committed:
remove files removed from scidoc4SRW3 branch

grantfirl committed:
remove one more file that was missed

grantfirl committed:
Merge pull request #263 from grantfirl/scidoc4SRW3_gjf

Scientific documentation for SRW v3.0.0 release (manually rebased)

MicroTed committed:
module_mp_nssl_2mom.F90 :
   - Fix bug for irenuc=2 (current default is irenuc=5, so no effect)
   - Some fixes for single-moment options (without hail and without graupel density)

grantfirl committed:
Merge branch 'ufs/dev' into nssl3m-update2025a

grantfirl committed:
Merge branch 'release/srw-v3' into srwv3-to-ufsdev

grantfirl committed:
Merge branch 'srwv3-to-ufsdev' into nssl3m-update2025a

grantfirl commented:
Note to any additional reviewers:

Changes in https://github.com/ufs-community/ccpp-physics/pull/254/commits/37405ee7321fc097b757ac42a940dd76b088f76e are scientific documentation changes from the SRW v3.0.0 release. They're being combined into this PR for testing/queue convenience. The documentation changes have been tested using UFS RTs in compile-only mode and should not otherwise affect the testing of the actual physics changes from the NSSL MP scheme from @MicroTed .

jkbk2004 commented:
All test is done at https://github.com/ufs-community/ufs-weather-model/pull/2661. @grantfirl @rhaesung @dustinswales can you merge this pr?
